zerver/tests/test_auth_backends.py

```diff
@@ -7212,5 +7216,156 @@ def test_invalid_realm_for_user_failure(self) -> None:
+        ), self.assertLogs("zulip.ldap", "DEBUG") as zulip_ldap_log:
+            self.assertFalse(
+                UserGroup.objects.filter(realm=get_realm("zulip"), name="cool_test_group").exists()
+            )
+
+            # Create group cool_test_group since it doesn't yet exist in Zulip
+            ensure_user_groups_for_ldap_sync_exist()
+            # Run the function a second time to test for bugs, this should noop at this point.
+            ensure_user_groups_for_ldap_sync_exist()
+
+            self.assertTrue(
+                UserGroup.objects.filter(realm=get_realm("zulip"), name="cool_test_group").exists()
+            )
+
+            user_group = UserGroup.objects.get(realm=get_realm("zulip"), name="cool_test_group")
+
+            self.assertFalse(
+                is_user_in_group(
+                    user_group,
+                    user_profile,
+                    direct_member_only=True,
+                )
+            )
+
+            sync_user_from_ldap(user_profile, mock.Mock())
+            self.assertTrue(
+                is_user_in_group(
+                    user_group,
+                    user_profile,
+                    direct_member_only=True,
+                )
+            )
+
+            # Add a user to a Zulip group that they are not member of in ldap.
+            # This implies that they should be deleted from the Zulip group
+            # upon the next sync.
+            user_profile = self.example_user("cordelia")
+            bulk_add_members_to_user_group(
+                user_group,
+                [user_profile.id],
+                acting_user=None,
+            )
+
+            self.assertTrue(
+                is_user_in_group(
+                    UserGroup.objects.get(realm=get_realm("zulip"), name="cool_test_group"),
+                    user_profile,
+                    direct_member_only=True,
+                )
+            )
+
+            # This should remove cordelia from cool_test_group
+            sync_user_from_ldap(user_profile, mock.Mock())
+
+            self.assertFalse(
+                is_user_in_group(
+                    UserGroup.objects.get(realm=get_realm("zulip"), name="cool_test_group"),
+                    user_profile,
+                    direct_member_only=True,
+                )
+            )
+
+        self.assertEqual(
+            zulip_ldap_log.output,
+            [
+                "DEBUG:zulip.ldap:creating zulip group cool_test_group from ldap for realm zulip",
+                "DEBUG:zulip.ldap:Syncing groups for user: 10",
+                "DEBUG:zulip.ldap:intended groups: {'cool_test_group'}; zulip groups: set()",
+                "DEBUG:zulip.ldap:add 10 to ['cool_test_group']",
+                "DEBUG:zulip.ldap:Syncing groups for user: 8",
+                "DEBUG:zulip.ldap:intended groups: set(); zulip groups: {'cool_test_group'}",
+                "DEBUG:zulip.ldap:removing groups {'cool_test_group'} from 8",
```

Comment:
Let's not hardcode the user IDs (8/1) -- you can use f-strings to substitute values like `hamlet.id`

Reply:
see 35798b2072fbf96b80aec032c52f341459ad84fd

(includes lint and  unit test fixes)